### PR TITLE
Make signal functions respect input dtype.

### PR DESCRIPTION
**Pull request by luigifcruz**
This pull request ensures that the data type from the input and output are the same. This results in an expressive 35% speedup in single-precision Hilbert operations.

```
## Upstream
%timeit sc.hilbert(data64)
126 ms ± 284 us per loop (mean ‡ std. dev. of 7 runs, 10 Loops each)
%timeit sc.hilbert(data32)
101 ms ± 662 us per loop (mean ‡ std. dev. of 7 runs, 10 Loops each)
```

```
## Patch
In [5]: %timeit sc.hilbert (data64)
130 ms ± 259 us per loop (mean + std. dev. of 7 runs, 10 Loops each)
In [6]: %timeit sc.hilbert(data32)
75.6 ms ± 174 us per Loop (mean ‡ std. dev. of 7 runs, 10 Loops each)
```

### CODE REVIEW

**Review by ilayn — APPROVED** [review]

File: scipy/signal/_signaltools.py
Comment (tylerjereddy) [inline scipy/signal/_signaltools.py:2343]: looks like it has core developer approval and a simple enough patch--should we worry that these two lines are not flushed by any tests?

image

Comment (larsoner) [reply scipy/signal/_signaltools.py:2343]: The only places `hilbert2` gets tested are with `assert_raises`:
```
(base) larsoner@bunk:~/python/scipy$ git grep hilbert2
...
scipy/signal/tests/test_signaltools.py:        assert_raises(ValueError, hilbert2, x)
scipy/signal/tests/test_signaltools.py:        assert_raises(ValueError, hilbert2, x)
scipy/signal/tests/test_signaltools.py:        assert_raises(ValueError, hilbert2, x, N=0)
scipy/signal/tests/test_signaltools.py:        assert_raises(ValueError, hilbert2, x, N=(2, 0))
scipy/signal/tests/test_signaltools.py:        assert_raises(ValueError, hilbert2, x, N=(2,))
```
@luigifcruz do you want to try adding some basic test in this PR? If not, I can push a quick commit to do it (assuming you have enabled "allow edits by maintainers")

Comment (luigifcruz) [reply scipy/signal/_signaltools.py:2343]: Sure @larsoner, I'll commit a test routine for this.

Comment (luigifcruz) [reply scipy/signal/_signaltools.py:2343]: Tests appear to be passing.

**Comment by luigifcruz** [comment]
I just noticed that the `lfilter_zi` function also doesn't respect the input type. The previous commit applies patches for this. A test for the function was also added. This commit might require re-review. cc @larsoner @ilayn

**Comment by luigifcruz** [comment]
The CI is now throwing three test errors after the last commit. I can't reproduce them in my local machine. I think this is related to the 32-bits environment. I'm not familiar with `RuntimeWarning: invalid value encountered in ? (vectorized)` error.

**Comment by luigifcruz** [comment]
I tried to reproduce the errors reported by the CI with a 32-bits Debian VM but no luck. This error appears to be similar to [this Satpy error](https://github.com/astropy/astropy/issues/12062).

**Comment by larsoner** [comment]
A restart fixed the errors, this LGTM, @tylerjereddy's comment has been addressed, and @ilayn approved already, so in it goes. Thanks @luigifcruz !

**Pull request merged**
